include/RAJA/exec-cuda/reduce_cuda.hxx

```diff
@@ -334,151 +493,202 @@ __device__ inline void atomicAdd(double *address, double value)
 
     int threadId = threadIdx.x + blockDim.x * threadIdx.y
                    + (blockDim.x * blockDim.y) * threadIdx.z;
 
     // initialize shared memory
+    T val = m_tally_device->tally;
     for (int i = BLOCK_SIZE / 2; i > 0; i /= 2) {
       // this descends all the way to 1
       if (threadId < i) {
-        // no need for __syncthreads()
-        sd[threadId + i] = m_reduced_val;
+        sd[threadId + i] = val;
       }
     }
+    if (threadId < 1) {
+      sd[threadId] = val;
+    }
     __syncthreads();
+#else
+    m_is_copy_host = true;
+    m_smem_offset = getCudaSharedmemOffset(m_myID, BLOCK_SIZE, sizeof(T));
+#endif
+  }
 
-    sd[threadId] = val;
-    __syncthreads();
+  /*!
+   * \brief Finish reduction on device and free memory on host.
+   *
+   * Destruction on host releases the device memory chunk for
+   * reduction id and id itself for others to use.
+   * Destruction on device completes the reduction.
+   *
+   * Note: destructor executes on both host and device.
+   */
+  __host__ __device__ ~ReduceMin<cuda_reduce<BLOCK_SIZE, Async>, T>()
+  {
+#if defined(__CUDA_ARCH__)
+    if (m_finish_reduction) {
+      extern __shared__ unsigned char sd_block[];
+      T *sd = reinterpret_cast<T *>(&sd_block[m_smem_offset]);
+
+      int threadId = threadIdx.x + blockDim.x * threadIdx.y
+                     + (blockDim.x * blockDim.y) * threadIdx.z;
 
-    for (int i = BLOCK_SIZE / 2; i >= WARP_SIZE; i /= 2) {
-      if (threadId < i) {
-        sd[threadId] = RAJA_MIN(sd[threadId], sd[threadId + i]);
-      }
       __syncthreads();
-    }
 
-    if (threadId < 16) {
-      sd[threadId] = RAJA_MIN(sd[threadId], sd[threadId + 16]);
-    }
-    __syncthreads();
+      for (int i = BLOCK_SIZE / 2; i >= WARP_SIZE; i /= 2) {
+        if (threadId < i) {
+          sd[threadId] = RAJA_MIN(sd[threadId], sd[threadId + i]);
+        }
+        __syncthreads();
+      }
 
-    if (threadId < 8) {
-      sd[threadId] = RAJA_MIN(sd[threadId], sd[threadId + 8]);
-    }
-    __syncthreads();
+      for (int i = WARP_SIZE / 2; i > 0; i /= 2) {
+        if (threadId < i) {
+          sd[threadId] = RAJA_MIN(sd[threadId], sd[threadId + i]);
+        }
+      }
 
-    if (threadId < 4) {
-      sd[threadId] = RAJA_MIN(sd[threadId], sd[threadId + 4]);
+      if (threadId < 1) {
+        _atomicMin<T>(&m_tally_device->tally, sd[threadId]);
+      }
     }
-    __syncthreads();
-
-    if (threadId < 2) {
-      sd[threadId] = RAJA_MIN(sd[threadId], sd[threadId + 2]);
+#else
+    if (!m_is_copy_host) {
+      releaseCudaReductionTallyBlock(m_myID);
+      releaseCudaReductionId(m_myID);
     }
-    __syncthreads();
+#endif
+  }
 
-    if (threadId < 1) {
-      sd[0] = RAJA_MIN(sd[0], sd[1]);
-      atomicMin(&(m_tallydata->tally), sd[0]);
+  /*!
+   * \brief Operator that returns reduced min value.
+   *
+   * Note: accessor only executes on host.
+   */
+  operator T()
+  {
+    if (Async) {
```

Comment:
So the ability to do this sort of thing is why the bool template parameter was added to the reduce policy, correct?